File: onedal/cluster/dbscan.cpp
/*******************************************************************************
* Copyright 2023 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

#include "oneapi/dal/algo/dbscan.hpp"

#include "onedal/common.hpp"
#include "onedal/version.hpp"

#include <regex>

namespace py = pybind11;

namespace oneapi::dal::python {

template <typename Task, typename Ops>
struct method2t {
    method2t(const Task& task, const Ops& ops) : ops(ops) {}

    template <typename Float>
    auto operator()(const py::dict& params) {
        using namespace dbscan;

        const auto method = params["method"].cast<std::string>();

        ONEDAL_PARAM_DISPATCH_VALUE(method, "brute_force", ops, Float, method::brute_force);
        ONEDAL_PARAM_DISPATCH_VALUE(method, "by_default", ops, Float, method::by_default);
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
    }

    Ops ops;
};

auto get_onedal_result_options(const py::dict& params) {
    using namespace dal::dbscan;

    auto result_options = params["result_options"].cast<std::string>();
    result_option_id onedal_options;

    try {
        std::regex re("\\w+");
        const std::sregex_iterator last{};
        const std::sregex_iterator first( //
            result_options.begin(),
            result_options.end(),
            re);

        for (std::sregex_iterator it = first; it != last; ++it) {
            std::smatch match = *it;
            if (match.str() == "responses") {
                onedal_options = onedal_options | result_options::responses;
            }
            else if (match.str() == "core_observation_indices") {
                onedal_options = onedal_options | result_options::core_observation_indices;
            }
            else if (match.str() == "core_observations") {
                onedal_options = onedal_options | result_options::core_observations;
            }
            else if (match.str() == "core_flags") {
                onedal_options = onedal_options | result_options::core_flags;
            }
            else
                ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(result_options);
        }
    }
    catch (std::regex_error& e) {
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(result_options);
    }

    return onedal_options;
}

struct params2desc {
    template <typename Float, typename Method, typename Task>
    auto operator()(const pybind11::dict& params) {
        using namespace dal::dbscan;

        const auto min_observations = params["min_observations"].cast<std::int64_t>();
        const auto epsilon = params["epsilon"].cast<double>();
        auto desc = descriptor<Float, Method, Task>(epsilon, min_observations);
        desc.set_mem_save_mode(params["mem_save_mode"].cast<bool>());
        desc.set_result_options(get_onedal_result_options(params));

        return desc;
    }
};

template <typename Policy, typename Task>
void init_compute_ops(py::module_& m) {
    m.def("compute",
          [](const Policy& policy,
             const py::dict& params,
             const table& data,
             const table& weights) {
              using namespace dbscan;
              using input_t = compute_input<Task>;

              compute_ops ops(policy, input_t{ data, weights }, params2desc{});
              return fptype2t{ method2t{ Task{}, ops } }(params);
          });
}

template <typename Task>
void init_compute_result(py::module_& m) {
    using namespace dbscan;
    using result_t = compute_result<Task>;

    py::class_<result_t>(m, "compute_result")
        .def(py::init())
        .DEF_ONEDAL_PY_PROPERTY(labels, result_t)
        .DEF_ONEDAL_PY_PROPERTY(modularity, result_t)
        .DEF_ONEDAL_PY_PROPERTY(community_count, result_t);
}

ONEDAL_PY_TYPE2STR(dbscan::task::clustering, "clustering");

ONEDAL_PY_DECLARE_INSTANTIATOR(init_compute_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_compute_result);

// TODO:
// change the name of modue for all algos -> cluster.
ONEDAL_PY_INIT_MODULE(dbscan) {
    using namespace dal::detail;
    using namespace dbscan;
    using namespace dal::dbscan;

    using task_list = types<task::clustering>;
    auto sub = m.def_submodule("dbscan");

#ifdef ONEDAL_DATA_PARALLEL_SPMD
    ONEDAL_PY_INSTANTIATE(init_compute_ops, sub, policy_spmd, task_list);
#else // ONEDAL_DATA_PARALLEL_SPMD
    ONEDAL_PY_INSTANTIATE(init_compute_ops, sub, policy_list, task_list);
    ONEDAL_PY_INSTANTIATE(init_compute_result, sub, task_list);
#endif // ONEDAL_DATA_PARALLEL_SPMD

}

} // namespace oneapi::dal::python

File: onedal/cluster/dbscan.py
# ===============================================================================
# Copyright 2023 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ===============================================================================

import numpy as np
from daal4py.sklearn._utils import get_dtype, make2d
from onedal._device_offload import supports_queue
from onedal.common._backend import bind_default_backend

from ..common._mixin import ClusterMixin
from ..datatypes import from_table, to_table
from ..utils.validation import _check_array


class DBSCAN(ClusterMixin):
    def __init__(
        self,
        eps=0.5,
        *,
        min_samples=5,
        metric="euclidean",
        metric_params=None,
        algorithm="auto",
        leaf_size=30,
        p=None,
        n_jobs=None,
    ):
        self.eps = eps
        self.min_samples = min_samples
        self.metric = metric
        self.metric_params = metric_params
        self.algorithm = algorithm
        self.leaf_size = leaf_size
        self.p = p
        self.n_jobs = n_jobs

    @bind_default_backend("dbscan.clustering")
    def compute(self, params, data_table, weights_table): ...

    def _get_onedal_params(self, dtype=np.float32):
        return {
            "fptype": dtype,
            "method": "by_default",
            "min_observations": int(self.min_samples),
            "epsilon": float(self.eps),
            "mem_save_mode": False,
            "result_options": "core_observation_indices|responses",
        }

    @supports_queue
    def fit(self, X, y=None, sample_weight=None, queue=None):
        X = _check_array(X, accept_sparse="csr", dtype=[np.float64, np.float32])
        sample_weight = make2d(sample_weight) if sample_weight is not None else None
        X_table, sample_weight_table = to_table(X, sample_weight, queue=queue)

        params = self._get_onedal_params(X_table.dtype)
        result = self.compute(params, X_table, sample_weight_table)

        self.labels_ = from_table(result.responses).ravel()
        if result.core_observation_indices is not None:
            self.core_sample_indices_ = from_table(
                result.core_observation_indices
            ).ravel()
        else:
            self.core_sample_indices_ = np.array([], dtype=np.intc)
        self.components_ = np.take(X, self.core_sample_indices_, axis=0)
        self.n_features_in_ = X.shape[1]
        return self
